[examples] Added error handling to the trib app for models and templates for errors and when no results are returned.

zhouyaoji opened the pull request:
The models of the mojits Blog, Calendar, Gallery, Github, and Youtube would occasionally not return results because of an issue with YQL or the original data source. Because the models didn't handle this case, the app would fail. Two shared templates were added that mojits could use to render HTML when there was an error or no results.
In addition, an error in the Calendar model that incorrectly parsed a string to extract a date was corrected.

Commit:
Added mocked data to the models for mojits that are making calls to 3rd party Web services.

Commit:
Merge branch 'develop' into trib_app_error_handling

Commit:
Added files using mock selector.

Commit:
Added templates for errors and no results.

Commit:
Added error handling for cases where no results are returned. Now there are shared templates for the error page and the case where there are no results.

Commit:
Deleted templates for errors and no results from Blog mojit. Those templates are now global.

Commit:
Merge branch 'develop' into trib_app_error_handling

Commit:
Removed selector  for now as an engineer is working on creating contextualized models that serve mocked data.

Commit:
Removed blackberry context.

zhouyaoji commented:
This PR is intended to fix issue #1253.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js:
```diff
@@ -38,12 +39,13 @@ YUI.add('Blog', function (Y, NAME) {
                     title: title,
                     results: data
-                });
+                }, data.view);
```
Models should not be telling controllers which view to select. Instead, models should return either results, or error, and controller should take appropriate action.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:
```diff
@@ -33,9 +34,14 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
             var results = [];
-            if (result.error === undefined) {
+            results.view = "index";
+            if (result.query && result.query.results === null) {
```
Only check for truthy or falsy

Commit:
Removed  template, had controllers handle error response, templates handle empty response.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js:
```diff
@@ -38,15 +39,21 @@ YUI.add('Blog', function (Y, NAME) {
-                    title: title,
-                    results: data
-                });
+                Y.log("In Blog controller: ", "info"); 
```
"info", NAME);

caridy commented on examples/developer-guide/dashboard/trib/mojits/Blog/controller.server.js:
```diff
@@ -38,15 +39,21 @@ YUI.add('Blog', function (Y, NAME) {
+                   // Error was found, so render `error` template.
+                   ac.done({ 
+                      title: title, 
+                      results: data 
```
do the `error` template requires `results`? or just the title?

caridy replied:
I will argue that the right thing here is to do ac.error(); and let the parent to deal with the error, but I might not have the full info here.

zhouyaoji replied:
The returned `data` object will have the results or an `error` object. The `error` template displays the error, which is not desired effect in production, but for an example app, it makes it easier to view errors. If you use `ac.error`, the child mojit's template will simply not be rendered, but the error will show up in the server logs. Just using `ac.error` might be the better solution though.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:
```diff
@@ -37,19 +38,17 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
 
                 if (result && result.query && result.query.results && result.query.results.item) {
                     results = result.query.results.item;
+                    Y.blogData = results;
```
we should NOT encourage people to do this kind of cache. If this cache is absolutely necesary, I will recommend do `YUI.Env.*`, which explicitly state that it is global.

zhouyaoji replied:
The caching implementation was created by the YUI team. I don't include it in the tutorial. Can you file a separate issue for this? Thanks.

imalberto replied:
@zhouyaoji You could cache the results by using a variable local to the module vs attaching it to the `Y` instance.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Blog/views/index.hb.html:
```diff
@@ -2,14 +2,17 @@
-            <ul>
-                {{#results}}
+            {{#each results}}
+                <ul> 
```
Each result will be wrap by an UL? sounds weird.

zhouyaoji replied:
The `index` template also handle the case where no results are shown; thus, I didn't want to have a `<ul>` if there were no results, but it shouldn't wrap each result with a `<ul>`. but I'll confirm and make changes if needed.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js:
```diff
@@ -30,17 +31,22 @@ YUI.add('Calendar', function (Y, NAME) {
+                if(data.error) {
+                    // Error was found: render `error` template.
+                    ac.done({ 
+                        title: title, 
```
same as above!

caridy commented on examples/developer-guide/dashboard/trib/mojits/Calendar/controller.server.js:
```diff
@@ -23,6 +23,7 @@ YUI.add('Calendar', function (Y, NAME) {
          */
         index: function (ac) {
             ac.models.get('CalendarModelYQL').getData({}, function (data) {
+                var title = "YUI Calendar Info";
```
Isn't this suppose to come from the lang bundle?

zhouyaoji replied:
The `lang` bundle is just used by the `PageLayout` mojit for the top-level heading as an example. The child mojits do not perform localization.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js:
```diff
@@ -27,6 +28,7 @@ YUI.add('CalendarModelYQL', function (Y, NAME) {
         },
         onDataReturn: function (cb, result) {
             Y.log("calendar.server onDataReturn called");
```
proper log signature across all these files please.

zhouyaoji replied:
Will do. Thanks!

caridy commented on examples/developer-guide/dashboard/trib/mojits/Calendar/models/calendar.server.js:
```diff
@@ -36,32 +38,25 @@ YUI.add('CalendarModelYQL', function (Y, NAME) {
+                        val.entry.summary.content = tempDate;
+                        Y.log(val.entry.summary.content);
+                    });
+                    Y.calendarData = results;
```
same as above.

caridy commented on examples/developer-guide/dashboard/trib/mojits/Calendar/views/index.hb.html:
```diff
@@ -2,11 +2,13 @@
-            <ul>
-                {{#results}}
+            {{#each results}}
+                <ul>
```
same as above

caridy commented:
Ok, I added a lot of comments, but I guess I got confused since that was not the original purpose of the PR, although some of them are relevant for the error manipulation. Feel free to ignore the others.

zhouyaoji commented:
@Caridy Thanks for your comments! I'll make the log statements more consistent, confirm that I'm not wrapping each result in a `<ul>` element, and have the children use an `ac.error` instead of using an `error` template. I'm going to leave the caching mechanism in place. We should file an issue to correct that.

Commit:
Removed  template, fixed tests, corrected templates.

Commit:
Added NAME to Y.log statement.

zhouyaoji commented:
@caridy I removed the `error` template and used `ac.error` instead, which will just show the errors in the server logs. I also cleaned up the `Y.log` statements, and corrected the templates so that not each result is wrapped in `<ul>` tags--you were correct as usual. Anyway, let me know if I can go ahead and merge the changes. Again, we can file an issue to fix the caching next. Thanks again!

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js:
```diff
@@ -18,37 +19,32 @@ YUI.add('GalleryModelYQL', function (Y, NAME) {
 
-                    //Y.log("cookedQuery: " + cookedQuery);
-
+                    //Y.log("cookedQuery: " + cookedQuery, "info", NAME);
```
`if (!result.error) {` would work too

zhouyaoji replied:
Thanks, I went with your suggestion.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Gallery/models/gallery.server.js:
```diff
@@ -18,37 +19,32 @@ YUI.add('GalleryModelYQL', function (Y, NAME) {
+                    Y.galleryCacheTime = new Date().getTime();
+                } else {
+                    results = new Error({ error: { description: "The returned response was malformed." }}); 
                 }
```
Probably want to follow the node convention:
- if error: `cb(results.error)`
- if no error: `cb(null, results)`

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Youtube/models/youtube.server.js:
```diff
@@ -20,38 +20,30 @@ YUI.add('YoutubeModelYQL', function (Y, NAME) {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
Same as above. Follow node callback convention.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Youtube/controller.server.js:
```diff
@@ -43,20 +41,23 @@ YUI.add('Youtube', function (Y, NAME) {
+                    ac.error(data);
+                } else {
+                    // Create data structure from Web service response.
+                    res = youtubeMap(ac, data);
```
It's better not to pass `ac` to this function. This function is only iterating through the data. Makes the intent clearer.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Github/models/yql.server.js:
```diff
@@ -24,33 +23,31 @@ YUI.add('StatsModelYQL', function (Y, NAME) {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
`cb(err)` vs `cb(null, results)`.

imalberto commented on examples/developer-guide/dashboard/trib/mojits/Blog/models/blog.server.js:
```diff
@@ -16,45 +17,42 @@ YUI.add('BlogModelYQL', function (Y, NAME) {
-                cb(result.error);
+                results = result;
             }
+            cb(results);
```
Same here with passing the error as 1st arg, but if no error, the results should be the 2nd arg.

imalberto commented:
Done with review.

Commit:
Merge branch 'develop' into trib_app_error_handling

Commit:
Made callback functions from models take an err parameter as is standard with Node.js.

Commit:
Removed the ActionContext object from mapping functions.

imalberto commented:
@zhouyaoji LGTM. Please check the build failures on Travis that it's not related to the PR. Otherwise +1
